Improve vSphere metric gathering

mgaruccio opened the pull request:
Please label this PR with one of the following labels, depending on the scope of your change:

- Enhancement


Please explain:

- WHAT: Continued using the govmomi library, but moved from using the `quickStats` property on the vm object and instead doing a query of the full set of real-time performance metrics.  Also captures some additional metadata about the VM and it's current state that is generally useful for most vSphere admins.  A few notes about the approach below
  - I added a setting for what the lowest-acceptable interval for the metrics should be and defaulted it to 1, that's a typical config and will prevent each polling cycle from hitting the vSphere database as the "real-time" metrics at that interval are kept in memory.  An enhancement I would like to make is to only get the longer-interval stats when they are actually updated and merely send a cached version at every new poll so the beat doesn't hit the DB every 20s for a metric that will only change every 5-60 minutes
  - I'm doing a lookup of resource pool information at the start of each polling cycle, which is an improvement over the existing design of doing that lookup on each VM as is curently done for network names and host information, but is ultimately still going to cause load issues on large vCenters, hoping for some feedback on how to cache that information so it only needs do be looked up when the beat starts up
  - also added some selectors for resource pool and datacenters which allow for limiting the data collection to only machines in those containers.  This is extremely messy but also the most efficient way I was able to work out actually doing that retrieval, happy to discuss/change if there is a better approach
  - the new `MaxQuerySize` parameter is needed as vSphere will only return a subset of VM's for a large query, by default 256 but that is a configurable stat.  
- WHY:  This is primarily meant to enable us to use the metricbeat module on our large vCenters and to gather the actually required stats.  

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

The module can be tested against the govcsim simulator, but I would appreciate some feedback or direction on how to get the actual tests in the test file working

## Related issues


- Relates: #18079 #6594 #20313 


- 

## Use cases

Expands the metrics gathering to the full set of vSphere metrics, enabling use of the elastic stack as a proper vmware monitoring platform.

## Other note

Worked with @SubhrataK this morning discussing the changes suggested here and she requested that the PR be submitted now so that we can more easily collaborate. I'm also available on the Elastic community Slack as "@Mike Garuccio"

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+		// Perform the query for all VMs at once
+		metricBase, err := perfManager.SampleByName(ctx, querySpec, metricsToQuery, chunk)
+		if err != nil {
+			fmt.Printf("Error retrieving metrics: %s\n", err)
+			return nil // this probably needs to be a different error
```
why not return error here?

mgaruccio replied:
Had this in place early in my exploration of editing the module and was not sure if returning error would have some adverse impact, updated.

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
-			},
-		}
+	// Create a map to easily look up counter names by ID
+	counterMap := make(map[int32]string)
```
prealloc the map?

shmsr commented on metricbeat/module/vsphere/_meta/config.reference.yml:
```diff
@@ -10,3 +12,7 @@
   # get_custom_fields: false
+  # Set a maximum query size when retrieving objects, this is typically 256 on vmvsphere 6.5 and later, default 256
+  # max_query_size: 256
+  # Set a maximum metric level to obtain, between 1 and 4, lower numbers are more granular, any number higher than one will hit the database rather than in-memory stats
```
```suggestion
  # Set a maximum metric level to obtain, between 1 and 4, lower numbers are more granular, and any number higher than one will hit the database rather than in-memory stats
```

mgaruccio replied:
fixed.

shmsr commented on metricbeat/module/vsphere/_meta/config.reference.yml:
```diff
@@ -1,7 +1,9 @@
   enabled: true
   metricsets: ["datastore", "host", "virtualmachine"]
-  period: 10s
+  # real-time stats are updated in vsphere every 20 seconds and are available at metric level 1 for vms and hosts:
+  # datastores are updated on the vsphere side every 5 minutes by default and shorter query times may compromise performance
```
```suggestion
  # Real-time stats are updated in vsphere every 20 seconds and are available at metric level 1 for vms and hosts:
  # datastores are updated on the vsphere side every 5 minutes by default and shorter query times may compromise performance
```

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+					event["datastore.id"] = datastoreIds
+				}
+
+				reporter.Event(mb.Event{
+					MetricSetFields: event,
+				})
```
Looks cleaner.

```suggestion
				reporter.Event(mb.Event{MetricSetFields: event})
```

mgaruccio replied:
thanks, I'm used to python/black and formatting options like that not really being a thing I can adjust, I'll make sure to pay more attention to that kind of thing moving forward when writing go.

shmsr replied:
you can use gofumpt. stricter gofmt (and gofmt compliant). it'd help (better than gofmt).

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -302,3 +385,185 @@ func getHostSystem(ctx context.Context, c *vim25.Client, ref types.ManagedObject
+}
+
+func (m *MetricSet) getVirtualMachineList(ctx context.Context, c *vim25.Client, mgr *view.Manager, includedPools []string) ([]mo.VirtualMachine, error) {
+
```
```suggestion
```

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -302,3 +385,185 @@ func getHostSystem(ctx context.Context, c *vim25.Client, ref types.ManagedObject
+
+	defer func() {
+		if err := v.Destroy(ctx); err != nil {
+			m.Logger().Debug(fmt.Errorf("error trying to destroy view from vshphere: %w", err))
```
no need to wrap the error here also using just Debugf here:
```suggestion
			m.Logger().Debugf("error trying to destroy view from vshphere: %s", err)
```

mgaruccio replied:
fixed.

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -302,3 +385,185 @@ func getHostSystem(ctx context.Context, c *vim25.Client, ref types.ManagedObject
+		return nil, fmt.Errorf("error in Retrieve: %w", err)
+	}
+
+	// skip filtering the list if no pools are specified
```
```suggestion
	// Skip filtering the list if no pools are specified
```

mgaruccio replied:
fixed

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -302,3 +385,185 @@ func getHostSystem(ctx context.Context, c *vim25.Client, ref types.ManagedObject
+	return rpMap, nil
+}
+
+// this is currently unused since it's a very expensive call, but if we could cache it it would be nice to have
```
```suggestion
// getVMInventoryPath is currently unused since it's a very expensive call, but if we could cache it it would be nice to have
```

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -302,3 +385,185 @@ func getHostSystem(ctx context.Context, c *vim25.Client, ref types.ManagedObject
+		if err != nil {
+			return "", err
+		}
+		pathElements = append([]string{entity.Name}, pathElements...)
```
any specific reason to not to do: `pathElements = append(pathElements, []string{entity.Name})`?

mgaruccio replied:
each iteration of the loop needs to add the new path element to the beginning of the slice so when they are joined into a string they make a correctly formatted folder path to the VM.  it would also work to instead append the name at the end of the slice and then reverse it before doing the join if that would be cleaner/fit the usual style guidelines better.

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
-			event.Put("cpu.free.mhz", freeCPU)
-		}
+	querySpec := types.PerfQuerySpec{
+		IntervalId: 20, // this likely needs to be made into a variable to support other-than-realtime metrics but it's not clear how to define that in the config and map to this and Level correctly
```
```suggestion
		IntervalId: 20, // NOTE: This likely needs to be made into a variable to support other-than-realtime metrics but it's not clear how to define that in the config and map to this and Level correctly
```

mgaruccio replied:
updated

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
-			} else {
-				if len(networkNames) > 0 {
-					event["network_names"] = networkNames
+		// assign metrics to the event output
```
```suggestion
		// Assign metrics to the event output
```

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
-					event["network_names"] = networkNames
+		// assign metrics to the event output
+		for _, base := range metricBase {
+			if metric, ok := base.(*types.PerfEntityMetric); ok {
```
anti pattern: see this example: https://www.jetbrains.com/help/resharper/InvertIf.html

Invert the condition and code would look cleaner.

mgaruccio replied:
appreciate the feedback on this one, got the code updated to invert the condition

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+		for _, base := range metricBase {
+			if metric, ok := base.(*types.PerfEntityMetric); ok {
+
+				// build initial even tobject with static values
```
```suggestion
				// Build initial even objects with static values
```

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+						counter := counterMap[series.Id.CounterId]
+						event.Put(counter, series.Value)
+					default:
+						m.Logger().Debug("Metric is of an unknown type, skipping")
```
```suggestion
						m.Logger().Debug("metric is of an unknown type, skipping")
```

In all other places you have use smallcase. So suggesting the change for consistency.

mgaruccio replied:
updated

shmsr reviewed (changes requested):
Haven't taken a look at the logic thoroughly yet but here are a few review comments.

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -302,3 +385,185 @@ func getHostSystem(ctx context.Context, c *vim25.Client, ref types.ManagedObject
+	return inventoryPath, nil
+}
+
+// contains checks if a slice contains a specific element.
+func contains(s []string, e string) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
```
So after 1.21 upgrade, we would remove the import package as slices.Contains will be part of stdlib.

mgaruccio replied:
thanks, I wasn't aware of that function or the fact that it exists in the go standard library, made the switch for all uses in this commit.

mgaruccio commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -302,3 +385,185 @@ func getHostSystem(ctx context.Context, c *vim25.Client, ref types.ManagedObject
+	// Traverse up the inventory tree to construct the full path
+	var pathElements []string
+	ref := vm.Reference()
+	for ref.Type != "Datacenter" {
```
Looping like this is the only way to walk up the entity tree and construct the full inventory path of the VM. each iteration is able to get the name of the previous entity's `Parent` and then another call needs to be made to get that entities `Parent` and so-on until you reach an entity of type `Datacenter` which will always be the top-level object in an inventory.  

needing to do this is really frustrating, but, since govmomi uses the SOAP api it's the only way to actually get this info, which is also why I'm not actually using this function in the code since it will generate as many as a few dozen api calls to retrieve.

mgaruccio commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+	}
+	var rpList []string
+	for _, rp := range rpMap {
+		rpList = append(rpList, rp.ref.Value)
```
done

mgaruccio commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+				}
+
+				if virtualMachine.Datastore != nil {
+					var datastoreIds []string
```
done

mgaruccio commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -302,3 +385,185 @@ func getHostSystem(ctx context.Context, c *vim25.Client, ref types.ManagedObject
+}
+
+func (m *MetricSet) getVmChunks(vms []mo.VirtualMachine) [][]types.ManagedObjectReference {
+	var vmsToQuery []types.ManagedObjectReference
```
done

mgaruccio commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,110 +123,174 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+				} else {
+					m.Logger().Debug("custom fields not activated or custom values not found/parse in Summary data. This " +
+						"is either a parsing error from vsphere library, an error trying to reach host/guest or incomplete " +
+						"information returned from host/guest")
```
updated

shmsr commented:
PR description and title were not in the recommended format. I have fixed them. PTAL.

mgaruccio commented:
worked through your comments on the PR and I think I got them all resolved.  I also realized that the items which require a lookup really should be cached and not looked up again unless needed.  Is there a standard way of doing that kind of thing within the overall project or examples I should look at?

mgaruccio commented:
Just pushed some additional code that makes a number of changes and additions.

The biggest shift is moving the `rpMap` to a property on the metricset object and only refreshing it when needed to save API calls.  I also do something similar with host information, and with gathering information about the datastore. 

I also shifted away from using additional calls at all to get network information and instead get that as part of the vm information gathering, which allows for getting the IP address information as well.

finally I added some additional disk information.

Also added the new fields that are now being collected to the fields.yml file (note that this is representative of the cluster I am working on, there will likely be other fields collected for vSphere clusters on different versions and/or hardware configurations)

shmsr commented on metricbeat/module/vsphere/virtualmachine/_meta/fields.yml:
```diff
@@ -4,60 +4,195 @@
+    - name: storage.total
+      type: long
+      description: "Total storage allocated to the virtual machine."
+    - name: storage.uncommited
```
spelling of uncommited is not correct

mgaruccio replied:
fixed

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,115 +130,277 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+				},
+				"storage": mapstr.M{
+					"committed":  virtualMachine.Summary.Storage.Committed,
+					"uncommited": virtualMachine.Summary.Storage.Uncommitted,
```
correct the spelling of uncommited
```suggestion
					"uncommitted": virtualMachine.Summary.Storage.Uncommitted,
```

mgaruccio replied:
fixed

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,115 +130,277 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+					datastore, ok := m.DsMap[ds.Value]
+					if !ok {
+						dsRefreshRequired = true
+						m.Logger().Debugf("datastore with id %s not found, will refresh datastore information at the end of the run")
```
```suggestion
						m.Logger().Debugf("datastore with id %s not found, will refresh datastore information at the end of the run")
```

used %s but no arg passed to Debugf?

mgaruccio replied:
not sure how I missed that, datastore value added to the debug statement

shmsr commented:
@mgaruccio Can you please sign this: https://www.elastic.co/contributor-agreement?

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,115 +130,277 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
 	mgr := view.NewManager(c)
 
-	v, err := mgr.CreateContainerView(ctx, c.ServiceContent.RootFolder, []string{"VirtualMachine"}, true)
+	// if perform initial population of the host, resource pool, network, and datastore maps
```
```suggestion
	// If perform initial population of the host, resource pool, network, and datastore maps
```

mgaruccio replied:
fixed formatting and clarified comment

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,115 +130,277 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+	}
+
+	// create bool variables that can be set to true if refreshes are needed
+	var rpRefreshRequired bool
+	var hostRefreshRequired bool
+	var dsRefreshRequired bool
```
```suggestion
	var rpRefreshRequired, hostRefreshRequired, dsRefreshRequired bool
```

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,115 +130,277 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+	// Create a Performance Manager instance
+	perfManager := performance.NewManager(c)
+
+	// gets a counterInfo object that describes all available counters for virtualMachines
```
```suggestion
	// Gets a counterInfo object that describes all available counters for virtualMachines
```

mgaruccio replied:
fixed

shmsr commented on metricbeat/module/vsphere/virtualmachine/virtualmachine.go:
```diff
@@ -104,115 +130,277 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
+		// Perform the query for all VMs at once
+		metricBase, err := perfManager.SampleByName(ctx, querySpec, metricsToQuery, chunk)
+		if err != nil {
+			return fmt.Errorf("Error retrieving metrics: %s\n", err)
```
```suggestion
			return fmt.Errorf("error retrieving metrics: %w", err)
```

mgaruccio replied:
fixed

ishleenk17 commented:
Thank you @mgaruccio for your contribution. This will definitely add value to the Vsphere Module.
Couple initial reviews (I will be doing detailed feature review shortly): 

1. The [asciidoc file](https://github.com/elastic/beats/blob/main/CHANGELOG.next.asciidoc) needs to be updated
2. The [data.json file ](https://github.com/elastic/beats/blob/main/metricbeat/module/vsphere/virtualmachine/_meta/data.json)should be updated with the latest data
3. Once I have reviewd the logic here, we need to add relevant tests to the [virtual machine test file](https://github.com/elastic/beats/blob/main/metricbeat/module/vsphere/virtualmachine/virtualmachine_test.go)
4. Please sign the CLA agreement.

mgaruccio commented:
> Thank you @mgaruccio for your contribution. This will definitely add value to the Vsphere Module. Couple initial reviews (I will be doing detailed feature review shortly):
> 
>     1. The [asciidoc file](https://github.com/elastic/beats/blob/main/CHANGELOG.next.asciidoc) needs to be updated
> 
>     2. The [data.json file ](https://github.com/elastic/beats/blob/main/metricbeat/module/vsphere/virtualmachine/_meta/data.json)should be updated with the latest data
> 
>     3. Once I have reviewd the logic here, we need to add relevant tests to the [virtual machine test file](https://github.com/elastic/beats/blob/main/metricbeat/module/vsphere/virtualmachine/virtualmachine_test.go)
> 
>     4. Please sign the CLA agreement.

Thanks, I'm working on getting the asciidoc and data.json file updates done now along with the couple of additional comments on the existing code.

This is going in on behalf of my employer and I'm working with them on getting the CLA signed (it won't be an issue, just needs to have legal take a quick look at it).

As for the tests, can you point me to any good resources on understanding how the existing tests work?  I spent a bit of time looking at them and couldn't quite wrap my head around how I should be structuring

mgaruccio commented:
Ok all requested updates done and the CLA has been signed.  I am working on the tests next.

One thing I did notice while setting up the data.json file that appeared to already have existed in the module previously is that the `event.duration` field is just counting up from the initialization of the beat.  Is that intended? and if not how should it be defined? as the actual duration the metrics represent (20 seconds by default) or the time in between polling cycles?

pierrehilbert commented:
As no updates since 2months, I'm putting back to draft for now.

erelling commented:
Please please complete this, is a sorely needed feature for us.
